### PR TITLE
INT-4265: Add Java DSL .routeByError() EIP-method

### PR DESCRIPTION
**Pull request by artembilan**
JIRA: https://jira.spring.io/browse/INT-4265

* Add `.routeByError()` EIP-method to the `IntegrationFlowDefinition`
based on the `ErrorMessageExceptionTypeRouter`
* Add missed `dynamicChannelLimit()` option to the `RouterSpec`
* Rework `RouterSpec#RouterMappingProvider` into the `ContextRefreshedEvent`
phase initialization to be sure that dependent `MappingMessageRouterManagement`
is fully initialized before applying mapping conversions

### CODE REVIEW

File: spring-integration-core/src/main/java/org/springframework/integration/dsl/RouterSpec.java [hunk 1]
```diff
@@ -178,28 +198,31 @@ void addMapping(Object key, NamedComponent channel) {
...
-			}
-			for (Map.Entry<Object, NamedComponent> entry : this.mapping.entrySet()) {
-				Object key = entry.getKey();
-				String channelKey;
-				if (key instanceof String) {
-					channelKey = (String) key;
-				}
-				else if (key instanceof Class) {
-					channelKey = ((Class<?>) key).getName();
+		public void onApplicationEvent(ContextRefreshedEvent event) {
```
Comment (garyrussell) [inline spring-integration-core/src/main/java/org/springframework/integration/dsl/RouterSpec.java]: Shouldn't we confirm that the event context is this bean's declaring context?

File: spring-integration-core/src/main/java/org/springframework/integration/dsl/IntegrationFlowDefinition.java [hunk 2]
```diff
@@ -2029,6 +2030,26 @@ public B routeToRecipients(Consumer<RecipientListRouterSpec> routerConfigurer) {
...
+	 *      .channelMapping(IllegalArgumentException.class, "illegalArgumentChannel")
+	 *      .subFlowMapping(MessageHandlingException.class, sf ->
+	 *                                sf.handle(...))
+	 *    )
+	 * }
+	 * </pre>
+	 * @param routerConfigurer the {@link Consumer} to provide {@link ErrorMessageExceptionTypeRouter} options.
+	 * @return the current {@link IntegrationFlowDefinition}.
+	 */
+	public B routeByError(
```
Comment (garyrussell) [inline spring-integration-core/src/main/java/org/springframework/integration/dsl/IntegrationFlowDefinition.java]: Perhaps `routeByThrowable` would be more correct?

The javadocs for `ErrorMessageExceptionTypeRouter` are incorrect too (`Exception` Vs. `Throwable`).

**Review by garyrussell — CHANGES_REQUESTED** [review]

File: spring-integration-core/src/main/java/org/springframework/integration/dsl/IntegrationFlowDefinition.java (hunk 2 above)
Comment (artembilan) [inline spring-integration-core/src/main/java/org/springframework/integration/dsl/IntegrationFlowDefinition.java]: Well, `ErrorMessage` accepts `Throwable`, too...
We don't pursue here the strong alignment with types we use. Just human-friendly is enough, IMO...

File: spring-integration-core/src/main/java/org/springframework/integration/dsl/RouterSpec.java (hunk 1 above)
Comment (artembilan) [inline spring-integration-core/src/main/java/org/springframework/integration/dsl/RouterSpec.java]: I think so. Will fix.
Thanks

File: spring-integration-core/src/main/java/org/springframework/integration/dsl/IntegrationFlowDefinition.java (hunk 2 above)
Comment (garyrussell) [inline spring-integration-core/src/main/java/org/springframework/integration/dsl/IntegrationFlowDefinition.java]: I agree, but specifically, here (and in the EMTR), we are routing on a `Throwable` payload, not an `ErrorMessage`.

Comment (garyrussell) [inline spring-integration-core/src/main/java/org/springframework/integration/dsl/IntegrationFlowDefinition.java]: `Error` has a meaning in Java and we shouldn't mislead.

**Comment by artembilan** [comment]
Pushed. But I renamed it into `routeByException` . Somehow I don't feel comfort with all those `.able()` in the public method API names...
If you insist, I'll rename to the `routeByThrowable()`. That is just IMHO...

**Comment by garyrussell** [comment]
Merged as f03356ccbcdf4394e7ab3daeeb3a7362a856e5f8

**Pull request closed without merging**